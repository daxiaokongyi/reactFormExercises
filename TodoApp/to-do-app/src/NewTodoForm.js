import React, {useState} from 'react';
import uuid from 'react-uuid';

const NewTodoForm = ({addTodo}) => {
    // used to reset the state
    const INITIAL_STATE = {
        todo : ""
    }

    // setup useState
    const [formData, setFormData] = useState({
        todo : ""
    });    

    // set handle submit
    const handleSubmit = evt => {
        evt.preventDefault();
        addTodo({...formData, id: uuid()});
        setFormData(INITIAL_STATE);
    }

    // set handle change
    const handleChange = evt => {
        const {name, value} = evt.target;
        setFormData(formData => ({...formData, [name]: value}));
    }

    return (
        <div>
            <form onSubmit={handleSubmit}>
                <label htmlFor="todo">New Todo: </label>
                <input
                    id="todo"
                    name="todo"
                    value={formData.todo}
                    onChange={handleChange}
                    type="text"
                />
                <button>Add a Todo</button>
            </form>
        </div>
    )
}

export default NewTodoForm;