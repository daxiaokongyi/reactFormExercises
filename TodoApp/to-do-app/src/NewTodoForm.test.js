import React from 'react';
import {fireEvent, render} from '@testing-library/react';
import NewTodoForm from './NewTodoForm';

// smoke test
test('renders without crashing', () => {
  render(<NewTodoForm/>);
});

// snapshot test
test('matches snapshots', () => {
  const {asFragment} = render(<NewTodoForm/>);
  expect(asFragment()).toMatchSnapshot();
})

// check create todo function on form submit
test("runs the create function on form submit", () => {
    const addMock = jest.fn();
    const {getByText} = render(<NewTodoForm addTodo={addMock}/>)
    const addBtn = getByText("Add a Todo");
    fireEvent.click(addBtn);
    expect(addMock).toHaveBeenCalled();
})

