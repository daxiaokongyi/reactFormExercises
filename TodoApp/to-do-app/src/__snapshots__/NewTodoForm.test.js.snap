// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="todo"
      >
        New Todo: 
      </label>
      <input
        id="todo"
        name="todo"
        type="text"
        value=""
      />
      <button>
        Add a Todo
      </button>
    </form>
  </div>
</DocumentFragment>
`;
